docs/versioned_docs/version-zenith/quickstart/292472-primitives.mdx

```diff
@@ -0,0 +1,71 @@
+
+# Quickstart
+
+<PartialExperimentalDocs />
+
+## Understand function primitives
+
+Module functions can accept arguments and return both basic and custom types. Dagger introduces valuable new primitives such as `Directory`, `Container`, and `Service`, and Dagger functions can use these as both inputs and outputs (in addition to your own custom types). This quickstart uses the `Directory` and `Container` primitives.
+
+### Strings
+
+To begin understanding this, call a function using the Dagger CLI:
+
+```shell
+dagger call -m github.com/kpenfound/dagger-modules/golang --proj https://github.com/dagger/dagger test --args ./core
+```
+
+This Go builder module exposes a `Test()` function that accepts a remote Go repository URL as input argument. It then runs the test suite in the repository and returns the test results to the caller. This example uses Dagger's GitHub repository as input and returns the results of a subset of Dagger's own test suite.
+
+Once the command completes, you should see output similar to the following:
+
+```shell
+✔ dagger call test [5m19.0s]
+┃ ok      github.com/dagger/dagger/core   0.397s
+```
+
+### Directories
+
+Dagger functions can also return directories. To see this in action, use the following command:
+
+```shell
+dagger call -m github.com/kpenfound/dagger-modules/golang --proj https://github.com/dagger/dagger build --args ./cmd/dagger
+```
+
+This example uses the same Go builder module as before, but this time calls its `Build()` function. Like the `Test()` function, the `Build()` function accepts a remote Go repository URL and builds a binary from the remote Go repository. It returns a Dagger `Directory` representing the state of the working directory in the repository after the build, similar to the output shown below:
+
+```
+✔ dagger call build [3m13.7s]
+┃ .changes
+┃ .git
+┃ .github
+...
+┃ sdk
+┃ telemetry
+┃ tracing
+```
+
+:::tip
+It's important to realize that even though you're testing and building a Go application in this example, you don't need to have Go, Git or any other local dependencies installed on the Dagger host. You only need the Dagger CLI and the ability to run containers. This is a very powerful feature of Dagger, because it allows development teams to create standardized tooling and eliminates all the variables and dependencies related to the host environment and/or configuration.
+:::
+
+### Containers
+
+Just as functions can return strings and directories, so too can they return containers. To see this in action, call another function using the Dagger CLI:
+
+```shell
+dagger call -m github.com/shykes/daggerverse/wolfi base container
+```
+
+This example uses a Wolfi container builder module. It returns a Dagger `Container` representing the built container. You should see the output below:
+
+```shell
+✔ dagger call base container [2.07s]
+┃ Container evaluated. Use "dagger call base container --help" to see available sub-commands.
+```
```

Comment:
```suggestion
### Pass a container as argument

In addition to basic types, Dagger defines powerful core types, which can be used in a function's arguments. For example, using the core `Container` type, a function can literally receive a container as argument. For example, a function that builds a container could use a default base image, but allow passing a custom base as an argument; or a custom publishing function could take a container as argument and upload it to a registry.

This feature is most powerful when calling a function from your own code (see [Create your first module](/428201/custom-modules/)). But you can also pass a container argument from the command-line. To do so, add the corresponding flag, followed by the address of an OCI image. The CLI will dynamically pull the image, and pass the resulting container object as argument to the function.

For example, let's call our `hello` function again, with a custom container in which to call the `figlet` tool. We must make sure that the container has the tool already installed.

```shell
dagger -m github.com/shykes/daggerverse/hello call hello --giant --figlet-container index.docker.io/hairyhenderson/figlet
```

You should see the image being pulled, then the following output:

```shell
✔ dagger call hello [1.67s]
┃  _   _      _ _                             _     _ _                                          
┃ | | | | ___| | | ___    __      _____  _ __| | __| | |                                         
┃ | |_| |/ _ \ | |/ _ \   \ \ /\ / / _ \| '__| |/ _` | |                                         
┃ |  _  |  __/ | | (_) |   \ V  V / (_) | |  | | (_| |_|                                         
┃ |_| |_|\___|_|_|\___( )   \_/\_/ \___/|_|  |_|\__,_(_)                                         
┃                     |/                                       
```

It worked! Let's see what happens if we pass a container that doesn't have the required tool installed:

```shell
dagger -m github.com/shykes/daggerverse/hello call hello --giant --figlet-container index.docker.io/alpine
```

You should see an error telling you that `figlet` could not be found:

```shell
✘ dagger call hello ERROR [0.98s]
[...]                                                     
✘ stdout ERROR [0.13s]
✘ exec figlet Hello, world! ERROR [0.11s]
┃ exec: "figlet": executable file not found in $PATH        
```